Dockerfile.builder

```diff
@@ -25,11 +25,16 @@ RUN apt-get update \
 		apt-utils \
 		binutils \
 		ca-certificates \
+		clang \
```

Comment:
@joestringer Shouldn't we specify the clang version to be the same one we use in production?

Reply:
I can see a few reasons you might want that:
1. At runtime, the compiler you use may affect operation of the datapath.
1. At datapath load time, the compiler (and loader) you use may affect whether you can load the programs or not, for example because different compiler optimizations are made.

For (1), that's what we have the end-to-end tests for. My intention behind this PR is for it to be used by some minor unit tests of the userspace golang side of the code. If something works, it's good enough for this use case.

For (2), we have specific ginkgo load complexity tests which test using the compiler we deploy, so we will not lose coverage by adding this here. The potential consequence would be that the unit tests fail to run, which we would catch when we start making use of this compiler in unit tests.

Overall I think ideally we should actually be testing multiple compilers rather than just testing one and presuming that others work. But that said, this is not the intention of this specific change.

IMO it's neither here nor there for how we select the compiler here, other than perhaps for visibility because it's not obvious which specific version is in use when these tests run. That said, this shouldn't be difficult to figure out when we need to know (eg, when a test fails specifically because of the compiler).

Do you have other thoughts or reservations about this?

Reply:
>Do you have other thoughts or reservations about this?

Yes, if we start testing against a different compiler than the one we use officially we will have different behavior for both cases. This can cause, among other things, tests passing with clang 8.0 but there will be a corner case where they will fail in a production environment for clang 3.9. If we start relying in e2e tests to catch bugs that were not caught in the unit tests running against clang 8.0 then why do we have the unit tests in the first place? Even if the unit test is basically comparing 2 bytes today we don't know what in the future we will add to those unit tests, we might start adding more complex tests where they will pass with clang 8.0, they will pass in e2e tests but they will fail in production environments.

Reply:
> If we start relying in e2e tests to catch bugs that were not caught in the unit tests running against clang 8.0 then why do we have the unit tests in the first place?

I've been looking at unit tests as a way to provide coverage on code that is simply not covered by other testing today, particularly on the userspace side of the interactions with the datapath. When it comes to the classes of bugs that may be introduced by differences to compiler versions, I see them as most likely falling under one of the following categories:

1. Clang optimizes the code differently and the programs simply won't load because of complexity exceeding the verifier limits. In this case, a simple load of the program that was built with the correct compiler should catch it. We have such tests already in Ginkgo.

1. Compiler bug that causes the logic to differ from what is intended by the code. This is always a possibility but given how much usage Clang gets, it is far less likely than other problems that we have few mitigations for, like developer error.

If there are other classes of bugs that could be introduced with different compiler versions, I'd be interested to hear them.

> we don't know what in the future we will add to those unit tests, we might start adding more complex tests where they will pass with clang 8.0, they will pass in e2e tests but they will fail in production environments.

That's a reasonable concern. That said, if we're not running e2e tests on expected production usage then I think we need to look a bit closer at our e2e tests. I see the point of the unit tests as ensuring that individual pieces of the functionality we provide each work as developers intended; The e2e tests should be about finding out whether the full deployment and integration works as intended, before someone deploys it. Unit tests should not be a substitute for a lack of e2e integration testing.

Reply:
>If there are other classes of bugs that could be introduced with different compiler versions, I'd be interested to hear them.

@joestringer In the past https://github.com/cilium/cilium/pull/505/commits/09820a079b68a817dd5723437a1d63be7a12b5e1

we added that check because a more recent clang version didn't run with an older kernel version. That's the sort of bugs we can't be sure they will never happen.

Reply:
@aanm thanks for the example, I agree that such bugs are impossible to guarantee they won't happen, other than by testing specific permutations of compiler+kernel. I see that by controlling the compiler version, it is easier to guarantee that our test environments have a limited set of permutations to go through so if we hit an issue then it should be a problem with the kernel. But that also explicitly says that we have zero visibility on this type of bug with other compilers because we don't even try to test different permutations.

If I follow the background of 09820a079b68a817dd5723437a1d63be7a12b5e1 correctly, when you say that it "didn't run" with an older kernel version, you mean that the Clang version compiled the datapath code differently and this caused the kernel to reject the program because Clang used features that are not available in the older kernel. Is that right? This type of bug is already picked up via runtime tests for the combination of compiler/kernel that we test with. No test coverage is provided for other combinations.

I believe that there's one possibility that I didn't cover in my previous response, which is that the BPF machine is updated with newer instructions and if Clang generates code using these instructions then the code cannot be loaded on older kernels. This type of issue tends to be handled these days via the Clang option `-mcpu=probe` where Clang attempts to detect the BPF instructions available at complie time, and generate the BPF instructions accordingly. There's already examples of this for subroutine support, and other BPF instruction changes are also being mulled at the moment, which again would require Clang changes and extension of the target BPF cpu version.

To me, this seems more like an argument why we should increase the number of compilers we test with so that we have a better understanding of compiler/kernel combinations. Why? Because I'm seeing this as a problem where specific combinations fail to load, and that is the problem we hit more often. So, if we test with more combinations, then we will be more likely to catch such problems during testing due to the diversity in the test environment. If, on the other hand, we think that it's more likely that we hit an issue where the datapath succeeds loading and has different behaviour at runtime, then that would be an argument for why we should lock the compiler version down for all testing.

Reply:
Follow-up from brief discussion during this morning's developer meeting, we hope to eliminate compilation pretty soon at which point this discussion becomes somewhat moot; we will ship bytecode compiled by a specific compiler version which would be generated at container image creation time.

Reply:
@joestringer Do you still want to push this as is?

Reply:
@aanm let me just rebase and re-run CI just in case there are any conflicts against master first, I'll push a new version shortly.